src/contracts/GPv2Settlement.sol

```diff
@@ -190,58 +190,63 @@ contract GPv2Settlement is ReentrancyGuard {
     /// @param clearingPrices An array of token clearing prices.
-    /// @param encodedTrades Encoded trades for signed orders.
+    /// @param trades Trades for signed orders.
     /// @return executedTrades Array of executed trades.
     function computeTradeExecutions(
         IERC20[] calldata tokens,
         uint256[] calldata clearingPrices,
-        bytes calldata encodedTrades
+        GPv2Trade.Data[] calldata trades
     ) internal returns (GPv2TradeExecution.Data[] memory executedTrades) {
-        (uint256 tradeCount, bytes calldata remainingEncodedTrades) =
-            encodedTrades.decodeTradeCount();
-        executedTrades = new GPv2TradeExecution.Data[](tradeCount);
-
-        GPv2Encoding.Trade memory trade;
-        uint256 i = 0;
-        while (remainingEncodedTrades.length != 0) {
-            remainingEncodedTrades = remainingEncodedTrades.decodeTrade(
+        GPv2Signing.RecoveredOrder memory recoveredOrder =
+            GPv2Signing.allocateRecoveredOrder();
+
+        executedTrades = new GPv2TradeExecution.Data[](trades.length);
+        for (uint256 i = 0; i < trades.length; i++) {
+            GPv2Trade.Data calldata trade = trades[i];
+
+            recoveredOrder.recoverOrderFromTrade(
                 domainSeparator,
                 tokens,
                 trade
             );
             computeTradeExecution(
-                trade,
+                recoveredOrder,
                 clearingPrices[trade.sellTokenIndex],
                 clearingPrices[trade.buyTokenIndex],
+                trade.executedAmount,
+                trade.feeDiscount,
                 executedTrades[i]
             );
-            i++;
         }
-
-        require(i == tradeCount, "GPv2: invalid trade encoding");
     }
 
     /// @dev Compute the in and out transfer amounts for a single trade.
     /// This function reverts if:
     /// - The order has expired
     /// - The order's limit price is not respected.
     ///
-    /// @param trade The trade to process.
+    /// @param recoveredOrder The recovered order to process.
     /// @param sellPrice The price of the order's sell token.
     /// @param buyPrice The price of the order's buy token.
+    /// @param executedAmount The portion of the order to execute. This will be
+    /// ignored for fill-or-kill orders.
+    /// @param feeDiscount The discount to apply to the final executed fees.
     /// @param executedTrade Memory location for computed executed trade data.
     function computeTradeExecution(
-        GPv2Encoding.Trade memory trade,
+        GPv2Signing.RecoveredOrder memory recoveredOrder,
         uint256 sellPrice,
         uint256 buyPrice,
+        uint256 executedAmount,
+        uint256 feeDiscount,
         GPv2TradeExecution.Data memory executedTrade
     ) internal {
-        GPv2Order.Data memory order = trade.order;
+        GPv2Order.Data memory order = recoveredOrder.data;
+        bytes memory orderUid = recoveredOrder.uid;
```

Comment:
Do the two extra input values make a difference? They are defined as `trade.executedAmount` and `trade.feeDiscount` anyway when the function is called, so you could simply define the two variables at the start of the function and then use these two variables.
```suggestion
        GPv2TradeExecution.Data memory executedTrade
    ) internal {
        GPv2Order.Data memory order = recoveredOrder.data;
        bytes memory orderUid = recoveredOrder.uid;
        uint256 executedAmount = trade.executedAmount;
        uint256 feeDiscount = trade.feeDiscount;
```

Reply:
I had something like this at some point :stuck_out_tongue:. My main issue with this is that there is a lot of ambiguity and implicit assumption:
- There is duplicate data, e.g. `order.receiver` and `trade.receiver`, which one should you use?
- Is the order already recovered from the trade or not?
- If we were to write a unit test for this method, there would be a bunch of duplicate and unused data (for example the trade signature doesn't matter or need to be valid, trade parameters don't need to match, etc.).